test: added coverage to mining_basic.py

kevkevinpal opened the pull request:
Included a test that checks if we call submitblock with block.vtx.empty() then it throws an rpc deserialization error, currently we only test if !block.vtx->IsCoinBase() throws an rpc deserialization error

I've tested to make sure this actually doing what I intended by breaking up this if block into two if blocks with different error messages and running the functional test
https://github.com/bitcoin/bitcoin/blob/322ec63b01499c1ec52d3912ee382ebd59f2366b/src/rpc/mining.cpp#L963

This change should increase the test coverage for the `submitblock()` rpc in `./src/rpc/mining.cpp`

theStack commented on test/functional/mining_basic.py:
```diff
@@ -127,6 +127,12 @@ def assert_submitblock(block, result_str_1, result_str_2=None):
         block.nBits = int(tmpl["bits"], 16)
         block.nNonce = 0
+
+        bad_block_empty_vtx = copy.deepcopy(block)
+
+        self.log.info("submitblock: Test block decode failure with empty vtx")
+        assert_raises_rpc_error(-22, "Block does not start with a coinbase", node.submitblock, bad_block_empty_vtx.serialize().hex())
```
nit: could just directly serialize a newly created temporary CBlock() instance, instead of creating a new variable
```suggestion
        self.log.info("submitblock: Test block decode failure with empty vtx")
        assert_raises_rpc_error(-22, "Block does not start with a coinbase", node.submitblock, CBlock().serialize().hex())
```
Would also suggest to move this down to the test checking the other condition for the same error message ("submitblock: Test invalid coinbase transaction").

kevkevinpal replied:
FIxed in [a7b46a1](https://github.com/bitcoin/bitcoin/pull/27603/commits/a7b46a1feae60e38fe4bdcacf5034f44cae49222)

Thanks for the suggestion!

theStack reviewed (commented):
Concept ACK

kevkevinpal committed:
test: added coverage to mining_basic.py

Included a test that checks if we call submitblock with
block.vtx.empty() then it throws an rpc deserialization error, currently
we only test if !block.vtx->IsCoinBase() throws an rpc deserialization
error

theStack reviewed (approved):
ACK a7b46a1feae60e38fe4bdcacf5034f44cae49222

sgtroy88 commented:
> Contributorbc1qjnpe7swnvwwkpfygrx5mtlp3vjjwg6kht8j7d5

sgtroy88 commented:
![_data_user_0_team opay pay_cache_image_manager_disk_cache_e9228085f80f410e13df5dcff2e1b77105e2f34d3568649ebdbc22ad6da75317 0](https://github.com/bitcoin/bitcoin/assets/42873100/6db8c102-94e0-446d-9eb9-8d5e250de9a3)
![Screenshot_20230604-035551](https://github.com/bitcoin/bitcoin/assets/42873100/8ee21d0a-8a69-4dcd-90a6-833b1c193674)
![Screenshot_20230604-032530](https://github.com/bitcoin/bitcoin/assets/42873100/606e7cd0-61da-4edc-873e-dd7a32ad8741)
